### ISSUE TITLE
Old value missing 

### ISSUE THREAD

**Issue by NiDHanWang** [zendesk/maxwell#1601]
Hi Team Maxwell:
 We're seeing a strange issue that the some of the  ‘old value’  in update data missed while many of them didn't even within the same table(Used ... to represent the content in ‘data’) :
No 'old data' :
`{"database":"start_formal_yousi","table":"yu_selling_value","type":"update","ts":1605867569,"xid":2896378,"xoffset":0,"data":{...}}`
With 'old data':
`{"database":"start_formal_yousi","table":"yu_selling_value","type":"update","ts":1605803432,"xid":249794049,"xoffset":0,"data":{...},"old":{...}}`
What could be the possible reason for this? @osheroff Thanks in advance!

**Comment by osheroff** [comment]
Hi @NiDHanWang, 
I don't know.  I've seen this bug report before and believe that it's true, I've just never gotten a reliable reproduction of it. 

can you do a `mysqlbinlog -vvv` at the binlog position of the update that's missing `old` data and report what you find here?  There's *one* possible explanation for the bug (which is just that the server didn't log the old data for various reasons), but I don't want to rule out a deeper issue....

**Comment by NiDHanWang** [comment]
Thanks for replying @osheroff !
We've confirmed with the DBA saying that they moved the mysql service to another server which casused the situation that some updated data had the 'old value' while many of them didn't .
Actually ,in the new server,all the 'old value' missed.
The mysql server variables 'binlog_row_image' in new server was set  to 'minimal' while in old server ,it's 'full'.We thought that might be the problem because if  'minimal' was set,server won't record the 'berfore image' of updated data according to mysql document.
But after we set the  'binlog_row_image'  to 'full',it still didn't work
Here's a JSON and it's binlog
JSON:
`{'data': {...},'database': 'start_formal_yousi', 'table': 'yu_selling_value', 'ts': 1606111165,'type': 'update','xid': 5665939,'xoffset': 0}`
Binlog:
![image](https://user-images.githubusercontent.com/54543492/100051617-4afc1900-2e57-11eb-81ec-6417ae979c9b.png)

**Comment by osheroff** [comment]
hi, can you copy/paste the text of these rather than screenshots? I would like to look at the binlog data as binary and see if there's a reason it might not be outputting `old` data, but I don't want to have to type everything in....

**Comment by NiDHanWang** [comment]
My bad (¬‿¬),here's the text

**JSON without old data:**
{'data': {...},'database': 'start_formal_yousi', 'table': 'yu_selling_value', 'ts': 1606111165,'type': 'update','xid': 5665939,'xoffset': 0}
**Binlog without old data:**

``` 
# at 234
#201123 13:01:59 server id 107339  end_log_pos 123 CRC32 0x29b731f7     Start: binlog v 4, server v 5.7.18-txsql-log created 201123 13:01:59
BINLOG '
R0K7Xw9LowEAdwAAAHsAAAAAAAQANS43LjE4LXR4c3FsLWxvZwAAAAAAAAAAAAAAAAAAAAAAAAAA
AAAAAAAAAAAAAAAAAAAAAAAAEzgNAAgAEgAEBAQEEgAAXwAEGggAAAAICAgCAAAACgoKKioAEjQA
Afcxtyk=
'/*!*/;
# at 9105505
#201123 13:59:25 server id 107339  end_log_pos 9105602 CRC32 0x43810323         Table_map: `start_formal_yousi`.`yu_selling_value` mapped to number 776
# at 9105602
#201123 13:59:25 server id 107339  end_log_pos 9105798 CRC32 0xd26ec104         Update_rows: table id 776 flags: STMT_END_F

BINLOG '
vU+7XxNLowEAYQAAAMLwigAAAAgDAAAAAAEAEnN0YXJ0X2Zvcm1hbF95b3VzaQAQeXVfc2VsbGlu
Z192YWx1ZQAQ/gMD/gMDAwMDAQEBEhIDDwj+Uf54AAA8AAAAIwOBQw==
vU+7Xx9LowEAxAAAAIbxigAAAAgDAAAAAAEAAgAQ/////wAAFDIwMDkwNDIxMDMzMDA0Nzc3NTYx
AAAAAIijAAADMjAweAAAAEgAAACILAAAsDYAAAAAAAABAACZp0lWbpmnSVZuRwIAAAblkLTnlJwA
ABQyMDA5MDQyMTAzMzAwNDc3NzU2MQAAAACIowAAAzIwMHgAAABJAAAAiCwAALA2AAAAAAAAAQAA
madJVm6Zp0lWbkcCAAAG5ZC055ScBMFu0g==
'/*!*/;
### UPDATE `start_formal_yousi`.`yu_selling_value`
### WHERE
###   @1='20090421033004777561' /* STRING(81) meta=65105 nullable=0 is_null=0 */
###   @2=0 /* INT meta=0 nullable=0 is_null=0 */
###   @3=41864 /* INT meta=0 nullable=0 is_null=0 */
###   @4='200' /* STRING(120) meta=65144 nullable=0 is_null=0 */
###   @5=120 /* INT meta=0 nullable=0 is_null=0 */
###   @6=72 /* INT meta=0 nullable=0 is_null=0 */
###   @7=11400 /* INT meta=0 nullable=0 is_null=0 */
###   @8=14000 /* INT meta=0 nullable=0 is_null=0 */
###   @9=0 /* INT meta=0 nullable=0 is_null=0 */
###   @10=1 /* TINYINT meta=0 nullable=0 is_null=0 */
###   @11=0 /* TINYINT meta=0 nullable=0 is_null=0 */
###   @12=0 /* TINYINT meta=0 nullable=0 is_null=0 */
###   @13='2020-09-04 21:25:46' /* DATETIME(0) meta=0 nullable=0 is_null=0 */
###   @14='2020-09-04 21:25:46' /* DATETIME(0) meta=0 nullable=0 is_null=0 */
###   @15=583 /* INT meta=0 nullable=0 is_null=0 */
###   @16='鍚寸敎' /* VARSTRING(60) meta=60 nullable=0 is_null=0 */
### SET
###   @1='20090421033004777561' /* STRING(81) meta=65105 nullable=0 is_null=0 */
###   @2=0 /* INT meta=0 nullable=0 is_null=0 */
###   @3=41864 /* INT meta=0 nullable=0 is_null=0 */
###   @4='200' /* STRING(120) meta=65144 nullable=0 is_null=0 */
###   @5=120 /* INT meta=0 nullable=0 is_null=0 */
###   @6=73 /* INT meta=0 nullable=0 is_null=0 */
###   @7=11400 /* INT meta=0 nullable=0 is_null=0 */
###   @8=14000 /* INT meta=0 nullable=0 is_null=0 */
###   @9=0 /* INT meta=0 nullable=0 is_null=0 */
###   @10=1 /* TINYINT meta=0 nullable=0 is_null=0 */
###   @11=0 /* TINYINT meta=0 nullable=0 is_null=0 */
###   @12=0 /* TINYINT meta=0 nullable=0 is_null=0 */
###   @13='2020-09-04 21:25:46' /* DATETIME(0) meta=0 nullable=0 is_null=0 */
###   @14='2020-09-04 21:25:46' /* DATETIME(0) meta=0 nullable=0 is_null=0 */
###   @15=583 /* INT meta=0 nullable=0 is_null=0 */
###   @16='鍚寸敎' /* VARSTRING(60) meta=60 nullable=0 is_null=0 */
SET @@SESSION.GTID_NEXT= 'AUTOMATIC' /* added by mysqlbinlog */ /*!*/;
DELIMITER ;
```






**JSON with old data:**
`{'commit': True,'data': {...},'database': 'start_formal_yousi','old': {'teacher_id': '20062123091793870667'},'table': 'yu_online_classes','ts': 1606119378,'type': 'update','xid': 5978563}`

**Binlog with old data:**
```
# at 234
#201123 13:01:59 server id 107339  end_log_pos 123 CRC32 0x29b731f7     Start: binlog v 4, server v 5.7.18-txsql-log created 201123 13:01:59
BINLOG '
R0K7Xw9LowEAdwAAAHsAAAAAAAQANS43LjE4LXR4c3FsLWxvZwAAAAAAAAAAAAAAAAAAAAAAAAAA
AAAAAAAAAAAAAAAAAAAAAAAAEzgNAAgAEgAEBAQEEgAAXwAEGggAAAAICAgCAAAACgoKKioAEjQA
Afcxtyk=
'/*!*/;
# at 98006541
#201123 16:15:36 server id 107339  end_log_pos 98006630 CRC32 0x887122de        Table_map: `start_formal_yousi`.`yu_online_classes` mapped to number 664
# at 98006630
#201123 16:15:36 server id 107339  end_log_pos 98006828 CRC32 0x504757ba        Update_rows: table id 664 flags: STMT_END_F

BINLOG '
qG+7XxNLowEAWQAAAGZ21wUAAJgCAAAAAAEAEnN0YXJ0X2Zvcm1hbF95b3VzaQAReXVfb25saW5l
X2NsYXNzZXMABv7+EhL+Dwr+PP48AAD+PDwAAN4icYg=
qG+7Xx9LowEAxgAAACx31wUAAJgCAAAAAAEAAgAG///AFDIwMDYyMTIzMDkxNzkzODcwNjY3FDIw
MDYxMTE4MTcwMjQyNjg5OTM3maac+OiZp+DV1xQyMDIwMDIxMzE4NTMyNDkzNDg3NgbmiJDpnZnA
FDIwMDQxMDE5MzExMDcxNTY3NTE1FDIwMDYxMTE4MTcwMjQyNjg5OTM3maac+OiZp+8D5BQyMDIw
MDIxMzE4NTMyNDkzNDg3NgbmiJDpnZm6V0dQ
'/*!*/;
### UPDATE `start_formal_yousi`.`yu_online_classes`
### WHERE
###   @1='20062123091793870667' /* STRING(60) meta=65084 nullable=0 is_null=0 */
###   @2='20061118170242689937' /* STRING(60) meta=65084 nullable=0 is_null=0 */
###   @3='2020-06-14 15:35:40' /* DATETIME(0) meta=0 nullable=0 is_null=0 */
###   @4='2020-11-16 13:23:23' /* DATETIME(0) meta=0 nullable=0 is_null=0 */
###   @5='20200213185324934876' /* STRING(60) meta=65084 nullable=0 is_null=0 */
###   @6='鎴愰潤' /* VARSTRING(60) meta=60 nullable=0 is_null=0 */
### SET
###   @1='20041019311071567515' /* STRING(60) meta=65084 nullable=0 is_null=0 */
###   @2='20061118170242689937' /* STRING(60) meta=65084 nullable=0 is_null=0 */
###   @3='2020-06-14 15:35:40' /* DATETIME(0) meta=0 nullable=0 is_null=0 */
###   @4='2020-11-23 16:15:36' /* DATETIME(0) meta=0 nullable=0 is_null=0 */
###   @5='20200213185324934876' /* STRING(60) meta=65084 nullable=0 is_null=0 */
###   @6='鎴愰潤' /* VARSTRING(60) meta=60 nullable=0 is_null=0 */
SET @@SESSION.GTID_NEXT= 'AUTOMATIC' /* added by mysqlbinlog */ /*!*/;
DELIMITER ;
# End of log file
/*!50003 SET COMPLETION_TYPE=@OLD_COMPLETION_TYPE*/;
/*!50530 SET @@SESSION.PSEUDO_SLAVE_MODE=0*/; 
```

**Comment by NiDHanWang** [comment]
Hi @osheroff ,we got the answer.

We read the binlog from replication instance. Normally, the instance should inherit the variables setting/change  from master automaticlly ,saying that the replica should have switched the **'binlog_row_image'** to **‘FULL’** after we changed it on Master.

But somehow it didn't.(The binlog we showed you is from Master,we cannot get the replica's binlog for some reason)

The problem was fixed after we restarted the replica.

Thanks for your replyings ;-)
